### PR TITLE
TYPE: New interactive inlay type hints support

### PR DESCRIPTION
**Pull request by ortem**
type_hints

### CODE REVIEW

**Comment by panstromek** [comment]
Wow, this is cool. Can you somehow use this without mouse? (because it seems you cannot move caret to typehint to do `ctrl+b`...)

**Comment by ortem** [comment]
@panstromek AFAIK it's impossible for now. I hope IntelliJ Platform will support such a feature someday!

File: src/192/main/kotlin/org/rust/ide/hints/RsPlainTypeHint.kt [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.hints
+
+import com.intellij.codeInsight.hints.InlayInfo
+import com.intellij.codeInsight.hints.Option
+import com.intellij.psi.PsiElement
+
+/** New interactive type hints are used instead */
```
Comment (mchernyavsky) [inline src/192/main/kotlin/org/rust/ide/hints/RsPlainTypeHint.kt]: Maybe add an (registry) option so that the user can switch between old and new type hints?

File: src/192/main/kotlin/org/rust/ide/hints/RsInlayTypeHintsProvider.kt [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+            }
+            """.trimIndent()
+
+    override fun createConfigurable(settings: Settings): ImmediateConfigurable = object : ImmediateConfigurable {
+        val showForVariables = "Show for variables"
+        val showForLambdas = "Show for closures"
+        val showForIterators = "Show for iterators"
+        val showObviousTypes = "Show obvious types"
+
+        private val varField = CheckBox(showForVariables)
```
Comment (mchernyavsky) [inline src/192/main/kotlin/org/rust/ide/hints/RsInlayTypeHintsProvider.kt]: Maybe move these settings to a separate PR?

File: src/192/main/kotlin/org/rust/ide/hints/RsTypeHintsPresentationFactory.kt [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+            listOf(traitPresentation, traitTypesPresentation).join()
+        }
+    }
+
+    private fun checkSize(level: Int, elementsCount: Int): Boolean =
+        level > FOLDING_LEVEL || elementsCount > FOLDING_ELEMENTS_COUNT
+
+    private fun isDefaultTypeParameter(argument: Ty, parameter: RsTypeParameter): Boolean {
+        val defaultType = parameter.typeReference?.type ?: return false
+        return argument.shortPresentableText == defaultType.shortPresentableText
```
[the remark this answers is not in the document]
Comment (ortem) [reply src/192/main/kotlin/org/rust/ide/hints/RsTypeHintsPresentationFactory.kt]: Sure, but I think we can ignore such cases not to use real name resolution here because type hints should be rendered as fast as possible.

File: src/192/main/kotlin/org/rust/ide/hints/RsInlayTypeHintsProvider.kt [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+    override fun createSettings(): Settings = Settings()
+
+    override fun getCollectorFor(file: PsiFile, editor: Editor, settings: Settings, sink: InlayHintsSink): InlayHintsCollector? =
+        object : FactoryInlayHintsCollector(editor) {
+
+            val typeHintsFactory = RsTypeHintsPresentationFactory(factory, settings.showObviousTypes)
+
+            override fun collect(element: PsiElement, editor: Editor, sink: InlayHintsSink): Boolean {
+                if (file.project.service<DumbService>().isDumb) return false
+                if (element !is RsElement) return false
```
[the remark this answers is not in the document]
Comment (ortem) [reply src/192/main/kotlin/org/rust/ide/hints/RsInlayTypeHintsProvider.kt]: I'm not sure there is a big difference because AFAIK this method only is invoked when `isDumb` is false (there are checks at call sites) and when the file's language is `RsLanguage`. I've just copied the sequence from another implementation of this method.

File: src/192/main/kotlin/org/rust/ide/hints/RsTypeHintsPresentationFactory.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (ortem) [reply src/192/main/kotlin/org/rust/ide/hints/RsTypeHintsPresentationFactory.kt]: Get rid of it. This heuristics will be introduced in later PR.

File: src/192/main/kotlin/org/rust/ide/hints/RsPlainTypeHint.kt (hunk 1 above)
Comment (ortem) [reply src/192/main/kotlin/org/rust/ide/hints/RsPlainTypeHint.kt]: Let's make sure that new type hints cover all the functionality of the old ones instead.

File: src/192/main/kotlin/org/rust/ide/hints/RsInlayTypeHintsProvider.kt (hunk 2 above)
Comment (ortem) [reply src/192/main/kotlin/org/rust/ide/hints/RsInlayTypeHintsProvider.kt]: These settings are copied from old hints so I think we should leave them in this PR

**Comment by mchernyavsky** [comment]
bors retry
